### DISCUSSION TITLE
How can I sort context menu items?

### DISCUSSION THREAD

**Discussion by bigplayer-ai** [moudey/Shell#132] [Q&A]
Hey Sometimes I get "keep on device" in other option cascade menu. I want to change it to be outside by default.
Is there an option to order/sort the the context menu items?

**Comment by moudey** [comment]
Moving child items to parent items is not currently supported.
Sorting is only available manually at the moment as well.

**Comment by bigplayer-ai** [comment]
How can I change it manually?

**Comment by moudey** [comment]
```
static
{
    item(find='"copy"'   pos=0) 
    item(find='"cut"'    pos=1) 
    item(find='"delete"' pos=2) 
    ...
}
```

**Reply by bigplayer-ai** [reply]
But I can't move child items to parent items this way?

**Reply by moudey** [reply]
This option not supported yet, It is only possible to move items to submenus

**Reply by Giresharu** [reply]
> ```
> static
> {
>     item(find='"copy"'   pos=0) 
>     item(find='"cut"'    pos=1) 
>     item(find='"delete"' pos=2) 
>     ...
> }
> ```

Hello! I am manually setting the order of items using `pos`, but the actual display effect is chaotic, as shown in the screenshot: 
![Snipaste_2025-01-08_04-19-52](https://github.com/user-attachments/assets/28e5b506-83e0-4a9a-8563-53ba7313bba9)


Here is a snippet of my code, which only inserts these lines at the beginning of the default modify.nss, without any changes to the rest of the content (these are items provided by qttabbar, and in the Chinese environment, it seems that I can only modify them using Chinese string characters).
```
modify(find="打开" image=icon.open pos=1)
modify(find="在新的标签中打开" image=icon.open_new_tab pos=2)
modify(find="在额外视图中打开"  pos=3)
modify(find="在新进程中打开"  pos=4)

modify(type="recyclebin" where=window.is_desktop and this.id==id.empty_recycle_bin pos=5 sep)
```

**Comment by RubicBG** [comment]
it's not "chaotic" - it just uses different logic
- modify(find="在新的标签中打开" image=icon.open_new_tab pos=2)
it counts 0, 1 2 , the position 2 is no 3rd place 
- modify(find="在额外视图中打开"  pos=3)
it counts 0, 1 2 3
- modify(find="在新进程中打开"  pos=4)
here is different.  it counts 0, 1 2 3 4, but it add separator before 4. Some system items goes with separator
- for the item on position 5 it counts as usual 0 1 2 3, but 4 is now the separator, not the moved item. 

what you can do is put them in the same position, in reverse order - so each subsequent one will shift the command down one position:
modify(find="在新进程中打开"  pos=1)
modify(find="在额外视图中打开"  pos=1)
modify(find="在新的标签中打开" image=icon.open_new_tab pos=1)
modify(find="打开" image=icon.open pos=1)

I hope it is clear enough that I find it difficult to give examples with Chinese item titles.

**Reply by Giresharu** [reply]
> it's not "chaotic" - it just uses different logic
> 
> * modify(find="在新的标签中打开" image=icon.open_new_tab pos=2)
>   it counts 0, 1 2 , the position 2 is no 3rd place
> * modify(find="在额外视图中打开"  pos=3)
>   it counts 0, 1 2 3
> * modify(find="在新进程中打开"  pos=4)
>   here is different.  it counts 0, 1 2 3 4, but it add separator before 4. Some system items goes with separator
> * for the item on position 5 it counts as usual 0 1 2 3, but 4 is now the separator, not the moved item.
> 
> what you can do is put them in the same position, in reverse order - so each subsequent one will shift the command down one position: modify(find="在新进程中打开" pos=1) modify(find="在额外视图中打开" pos=1) modify(find="在新的标签中打开" image=icon.open_new_tab pos=1) modify(find="打开" image=icon.open pos=1)
> 
> I hope it is clear enough that I find it difficult to give examples with Chinese item titles.

Thank you for your answer! I will try again! My Windows license does not allow me to switch the system language. If there are any issues next time, I will try to use IDs instead of strings to avoid asking in Chinese.

**Reply by Giresharu** [reply]
> it's not "chaotic" - it just uses different logic
> 
> * modify(find="在新的标签中打开" image=icon.open_new_tab pos=2)
>   it counts 0, 1 2 , the position 2 is no 3rd place
> * modify(find="在额外视图中打开"  pos=3)
>   it counts 0, 1 2 3
> * modify(find="在新进程中打开"  pos=4)
>   here is different.  it counts 0, 1 2 3 4, but it add separator before 4. Some system items goes with separator
> * for the item on position 5 it counts as usual 0 1 2 3, but 4 is now the separator, not the moved item.
> 
> what you can do is put them in the same position, in reverse order - so each subsequent one will shift the command down one position: modify(find="在新进程中打开" pos=1) modify(find="在额外视图中打开" pos=1) modify(find="在新的标签中打开" image=icon.open_new_tab pos=1) modify(find="打开" image=icon.open pos=1)
> 
> I hope it is clear enough that I find it difficult to give examples with Chinese item titles.

Hi! 

Today, I tried setting the positions of a series of items to 0, but found that they are not arranged in the order from top to bottom. I still don't understand why. 

I tried using the IDs, but there are many contents that can't be found in the IDs, so I had to use Chinese strings.  Here is all my code: 
```
settings
{
	priority=1
	exclude.where = !process.is_explorer
	showdelay = 200
	// Options to allow modification of system items
	modify.remove.duplicate=1
	tip.enabled=true
}

import 'imports/theme.nss'
import 'imports/images.nss'

// Remove irrelevant items
modify(mode=mode.multiple
    where=this.id(id.restore_previous_versions,id.cast_to_device)
    vis=vis.remove)
modify(find="使用火绒" vis=vis.remove)
modify(find="在额外视图中打开" vis=vis.remove)
modify(where=this.id==id.open_in_new_process vis=vis.remove)
modify(find="编辑" vis=vis.remove)
modify(find="在新标签页中打开" vis=vis.remove)
modify(find="在当前标签中打开" vis=vis.remove)

// Set pos to 0
modify(where=this.id==id.open image=icon.open_new_window pos=0)
modify(where=this.id==id.run_as_administrator pos=0)
modify(find="在新的标签中打开" image=icon.open_in_new_tab pos=0)
modify(find="在新窗口中打开" image=icon.open_new_window pos=0)
separator()
modify(where=this.id==id.open_with pos=0)
modify(find="添加到收藏夹" pos=0)
modify(find="7-Zip" title="解压缩" image=icon.compressed pos=0)
```

I've removed some irrelevant parts and moved some of the items' pos to 0. However, the order is still not in the expected order : 

The `解压缩` that should be at the end is now the first, the `id.open (打开)` that should be first is now second, and the `在新的标签中打开` (id.run_as_administrator is hidden when in a folder) that should be second is now fourth.
![Snipaste_2025-01-13_02-44-21](https://github.com/user-attachments/assets/36981f5b-ac40-4bb6-afcb-92cdbf4e9926)

I really can't figure out why some items have such a high priority, being set to 0 or top makes them higher than others with the same settings. Is there some hidden setting at play here?

To meet expectations, I have to forcibly set some items with particularly high priority to pos=1, and each time I encounter the same issue, I have to sequentially lower their pos. However, this approach leads to a significant problem, as some items that have not been modified may abruptly intersperse among these pos. I cannot simply sort some items in order at the top to separate them from these unmodified items (the same applies to the bottom). I can only write down all the items I see one by one. When I want to place an item in order at the top and have written it as pos 7, I need to modify all the items in the middle to pos 8 or below; otherwise, they may be inserted into the items I expect to be at the top, which are above pos 7.

**Reply by RubicBG** [reply]
i will try to make it clearer:
- The most important thing to understand is that position is ALWAYS relative.
```
item D
item C
item A
item B
```
if you move `item D` down from pos=0 to pos=4, the position of D is 3 - `item C` is 0, `item A` is 1, `item B` is 2 and `item D` is 3
if you move `item D` down from pos=0 to pos=3, it goes before `item B` and the position is 2
Every second modify() counts from a "new arrangement"- in you case the 9th item with pos=8 has 9 times "new arrangements" before it
- an existing (system) separator is count, but modify() does not work gut with added (custom) separator like: 
```
modify(...)
separator()  // this do "nothing", maybe add separator first before every modify, i don't remember   
modify(...)
```
- Usually the first item is at position 0, but before 0 there may exist "Top" (and accordingly "Bottom"). This is not a "numerical number"
https://nilesoft.org/docs/configuration/properties#position

**Reply by Giresharu** [reply]
> i will try to make it clearer:
> 
> * The most important thing to understand is that position is ALWAYS relative.
> 
> ```
> item D
> item C
> item A
> item B
> ```
> 
> if you move `item D` down from pos=0 to pos=4, the position of D is 3 - `item C` is 0, `item A` is 1, `item B` is 2 and `item D` is 3 if you move `item D` down from pos=0 to pos=3, it goes before `item B` and the position is 2 Every second modify() counts from a "new arrangement"- in you case the 9th item with pos=8 has 9 times "new arrangements" before it
> 
> * an existing (system) separator is count, but modify() does not work gut with added (custom) separator like:
> 
> ```
> modify(...)
> separator()  // this do "nothing", maybe add separator first before every modify, i don't remember   
> modify(...)
> ```
> 
> * Usually the first item is at position 0, but before 0 there may exist "Top" (and accordingly "Bottom"). This is not a "numerical number"
>   [nilesoft.org/docs/configuration/properties#position](https://nilesoft.org/docs/configuration/properties#position)

I'm sorry, I don't understand what it means to `move from pos=0 to pos=4` . If I set all items to pos=0, why would their order change?

**Reply by RubicBG** [reply]
```js
// Remove irrelevant items (compact version)
remove(mode=mode.multiple where=this.id(id.open_in_new_process, id.restore_previous_versions, id.cast_to_device))
remove(find="使用火绒|在额外视图中打开|编辑|在新标签页中打开|在当前标签中打开")
// use reverse order - shift one by one down to avoid index changes
modify(pos='top' find="添加到收藏夹")
modify(pos='top' where=this.id==id.open_with)
modify(pos='top' find="在新窗口中打开" image=icon.open_new_window sep='after')
modify(pos='top' find="在新的标签中打开" image=icon.open_in_new_tab)
modify(pos='top' where=this.id==id.run_as_administrator)
modify(pos='top' where=this.id==id.open image=icon.open_new_window)
// use indexof to move to position base on already existing item
modify(pos=indexof("添加到收藏夹", 1) find="7-Zip" title="解压缩" image=icon.compressed)
```
i cant test it

**Reply by RubicBG** [reply]
```
item D // pos 0
item C // pos 1
item A // pos 2
item B // pos 3
```
let say you want them in order - A B C D
modify(find = 'item D' pos=4) - to go to the end, the new order is 
```
item C // pos 0
item A // pos 1
item B // pos 2
item D // pos 3
```
modify(find = 'item c' pos=3) - to go to the end, the new order is 
```
item A // pos 0
item B // pos 1
item C // pos 2
item D // pos 3
```
every modify() check the "new order" first!

**Reply by Giresharu** [reply]
> ```
> item D // pos 0
> item C // pos 1
> item A // pos 2
> item B // pos 3
> ```
> 
> let say you want them in order - A B C D modify(find = 'item D' pos=4) - to go to the end, the new order is
> 
> ```
> item C // pos 0
> item A // pos 1
> item B // pos 2
> item D // pos 3
> ```
> 
> modify(find = 'item c' pos=3) - to go to the end, the new order is
> 
> ```
> item A // pos 0
> item B // pos 1
> item C // pos 2
> item D // pos 3
> ```
> 
> every modify() check the "new order" first!

Does this mean that I need to know the original position of an unmodified item? The items that pop up when right-clicking on different file types vary; should I compile a list of all possible items and assign a unique position to each item in order, considering that any same position could potentially lead to unexpected sorting?

**Reply by RubicBG** [reply]
A "perfect" organized main menu is impossible - even Windows can't do it right. You can reach a level that satisfies you. Other methods of arrangement were discussed in discord, but whichever method you choose, there will always be something to rearrange.

**Reply by Giresharu** [reply]
> A "perfect" organized main menu is impossible - even Windows can't do it right. You can reach a level that satisfies you. Other methods of arrangement were discussed in discord, but whichever method you choose, there will always be something to rearrange.

I searched on Discord and saw a similar situation where an indexof is forcibly used to set a relative pos. However, it still seems to fail in achieving a comfortable sorting, because the item used for indexof is not always present. Once it's missing, the item using indexof gets pushed to the last position. I've tried many methods and found that only by manually copying out all possible items' name and modifying each one's absolute position can an abrupt sorting be avoided.

But I'm not sure if there's a way to print out a list of all possible items. Trying each file type one by one would be too silly.

**Reply by RubicBG** [reply]
- indexof has three parameters: `indexof('find', 'if_found_move_by', 'if_not_found_default_pos'='bottom')`, but the third parameter does not work at the moment. 
- in theory every windows system command can be replace with NS command, after that sorting will be easy, but for every command the conditions to show up must be set and that is not so easy task

If I find a method, I will let everyone know.
